### PR TITLE
[react-navigation] route params, headerBackTitle fixups or so.

### PR DESCRIPTION
**Pull request by phanalpha**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://reactnavigation.org/docs/navigators/stack#headerBackTitle>>

### CODE REVIEW

**Review by huhuanming — APPROVED** [review]
LGTM

File: types/react-navigation/index.d.ts
Comment (abrahambotros) [inline types/react-navigation/index.d.ts]: Are you sure? I think `params` **is** empty when not given?

File: types/react-navigation/index.d.ts
Comment (abrahambotros) [inline types/react-navigation/index.d.ts]: This is typed as just `string` in `TypeDefinition.js`. Is there a specific reason to make this allow `null` too? Just wondering, since this introduces a difference from the official Flow typings.

File: types/react-navigation/index.d.ts
Comment (abrahambotros) [inline types/react-navigation/index.d.ts]: Can you move all of this up into the `CUSTOM CONVENIENCE INTERFACES` area (between the `BEGIN` and `END` marks)? Maybe put this next to the stack navigator stuff in that area?

**Review by abrahambotros — CHANGES_REQUESTED** [review]
Some questions and suggestions. Also, can you add tests? In addition to general resilience from testing, would be great to show exactly how/where this would be used.

File: types/react-navigation/index.d.ts
Comment (phanalpha) [inline types/react-navigation/index.d.ts]: I think it is the user's responsibility to decide whether `params` could be optional. It would be a logical error if the `params` is required but not provided. Mark `params` always optional force an extra check before use it, which is inconvenient and seems unnecessary.

File: types/react-navigation/index.d.ts
Comment (phanalpha) [inline types/react-navigation/index.d.ts]: It is described in the [document](https://reactnavigation.org/docs/navigators/stack#headerBackTitle) and it does work. [Maybe types](https://flow.org/en/docs/types/maybe/) accept the provided type as well as null or undefined. Typescript do not accept null unless specified.

File: types/react-navigation/index.d.ts
Comment (phanalpha) [inline types/react-navigation/index.d.ts]: `react-navigation.js` exports views and header (`Header`, `HeaderTitle`, `HeaderBackTitle`), most of which are not included in current ts definition. I'm not sure if it is better to have it's own section. And I wonder `CUSTOM CONVENIENCE INTERFACES` is customized by who, or just be distinguished with `TypeDefinition.js`.

File: types/react-navigation/index.d.ts
Comment (huhuanming) [inline types/react-navigation/index.d.ts]: In strict mode，It requires that declare null when value is optional.

a number of cases In `react-native.d.ts`.

Comment (abrahambotros) [inline types/react-navigation/index.d.ts]: @phanalpha Ah, sorry, I forgot `null` was a valid value to pass for `headerBackTitle` specifically. Thanks for correcting me, this looks good.

File: types/react-navigation/index.d.ts
Comment (abrahambotros) [inline types/react-navigation/index.d.ts]: Good point, that section naming is a bit dated now that other things have been put in there, so I agree that `CUSTOM CONVENIENCE INTERFACES` is a bit inaccurate. You can leave this as-is for now, and maybe we can reorganize everything in a separate PR later. Thoughts on labeling sections either just as `Header`, `StackNavigator`, etc., versus by corresponding `react-navigation`-repo file names (`src/views/HeaderBackButton`)?

For the former, that's pretty simple, but if we want to point to the `src/...` file in a comment, it will look kinda messy. On the other hand, for the latter, there might be cases where a definition doesn't quite fit under a single file name.

Just thinking out loud, not important for this PR, but open to any opinions now. Thanks!

File: types/react-navigation/index.d.ts
Comment (abrahambotros) [inline types/react-navigation/index.d.ts]: Hm, this seems wrong to me. If we're typing out something, we probably want it to be as accurate as possible, even if that isn't quite as convenient sometimes. Here, you're suggesting trading off accuracy for convenience it seems? I think that this library/typing should be generic and not make assumptions about its use - just stick as close to what the actual underlying types are at runtime as possible.

I do understand it's a bit inconvenient, but why not just cast each time you need to access `params`? For example, `const params: ThisComponentsNavigationParamsInterface = this.props.navigation.state.params`?

Lastly, I think the biggest point is that this can easily result in trying to access an undefined object in many cases. Consider the following example.

```
interface ThisComponentsNavigationParamsInterface {
    something?: string,
}

...

render() {
    if (this.props.navigation.state.params.something) {
       return ( <Text>...</Text> );
    } else {
       return ( <View>...</View> );
    }
}

...

<From some other component>
`this.props.navigation.navigate(ThisComponentRouteName);
```

This would result in a runtime exception when accessing `state.params.something`, since `params` is actually totally empty, NOT an empty object. So in this case you'd have to check for `state.params && state.params.something`, which then is as inconvenient as you were saying before, just in the other direction.

You could argue that this is still somewhat the user's fault, since maybe they should have allowed `something` to also be `null`, and then passed it in explicitly in the `navigate` call as `null`, but this seems like we're building up more gotchas than solving problems for the general user, especially if they're new.

Thoughts?

Comment (phanalpha) [inline types/react-navigation/index.d.ts]: I totally agree with "we probably want it to be as accurate as possible". If a screen expects its params to be some `T`, it should be free to express this, other than be forced to be `T | undefined`. As in typical react component situation, a component could expect its props type `P`. If an undefined is expected, it could be expressed as `T | undefined` explicitly. Although currently, this could not be enforced by the compiler as props do. I think it would be a logical error to provide an unexpected undefined (as params). To force the (undefined) check does not solve the problem, and make the boundary fuzzy.

Comment (abrahambotros) [inline types/react-navigation/index.d.ts]: Wow, so sorry, you're totally right. I was mistakenly thinking of this as still being `params?: NavigationParams`, not the typed-generic `Params` (which is even more ironic since I made this change not too long ago in https://github.com/phanalpha/DefinitelyTyped/commit/9be2563bfc81f9127474f0c319c54532a3321517). This should indeed be changed according to what you did, since this is a typed generic, and the user should in general be responsible for adding `undefined if they want. But if this were still the generic *map* `NavigationParams`, then we would have wanted to leave this as `params?:`.

Really sorry about that, I've done some really sloppy and rushed work reviewing your PR. But thanks for the patience and for keeping up the explanation of your approaches and changes!

The only thing here is that unfortunately I believe my example with the empty `state.params` could still be applicable if end-users don't take care. I've copied it below:

```
interface ThisComponentsNavigationParamsInterface {
    something?: string,
}

...

render() {
    if (this.props.navigation.state.params.something) {
       return ( <Text>...</Text> );
    } else {
       return ( <View>...</View> );
    }
}

...

<From some other component>
this.props.navigation.navigate(ThisComponentRouteName);
```

Technically, what should probably be done instead is:
```
const params: ThisComponentsNavigationParamsInterface = {};
this.props.navigation.navigate(ThisComponentRouteName, params);
```
But this may be a gotcha for new users who are more loosely using the typed, and who would still expect at least an empty object in `this.props.navigation.state.params`. Thoughts there?

**Comment by paulvanbrenk** [comment]
@abrahambotros looks like you still want some changes?

**Comment by abrahambotros** [comment]
@paulvanbrenk I think at this point @phanalpha has rebutted most of my concerns, with just a couple things we can still discuss in the comment threads (and perhaps asynchronously from merging). The only main change I'd still like to see is the addition of at least a couple small tests (in particular for the `HeaderBackButton`), or alternatively an explanation of why this isn't necessary/would be overkill. @phanalpha does this sound good?

**Review by abrahambotros — APPROVED** [review]
@phanalpha looks good, thanks for adding the tests, and apologies again for the messy review work earlier. I think this is ready to merge with the help of @paulvanbrenk or any other maintainers, unless anyone else has any objections.

File: types/react-navigation/index.d.ts
Comment (phanalpha) [inline types/react-navigation/index.d.ts]: I would prefer to type `navigate()` more specifically, say a series of `navigate('routeNameX', params: ParamX)` or so, so that the compiler could do all the checks. The navigation logic implies a screen (local) is mostly coupled with the navigator (global). An ideal type system might be too much complicated to implement currently if possible, I think.

**Comment by paulvanbrenk** [comment]
@phanalpha Thank, if you get the merge conflict fixed we can merge this in.

File: types/react-navigation/index.d.ts
Comment (abrahambotros) [inline types/react-navigation/index.d.ts]: Yeah, that would certainly be the best case, but as you mentioned, given the decoupling in typings between the local and global contexts, this seems intractable at the time for compilation steps (and anything else before runtime). We'll just leave it up to the user to type their own params at this time, with my suggestion being the form mentioned in my comment above, and copied below:

```
const params: ThisComponentsNavigationParamsInterface = {};
this.props.navigation.navigate(ThisComponentRouteName, params);
```

**Pull request merged**
